Implementation update

awilkins opened the pull request:
* Used Atom's `workspace.scan` instead of rolling own file traversal code
* Replaced path config with a single glob compatible with `workspace.scan`
* Implemented parameter insertion for multiple group matches
* Implemented matching for literal string matchers as well as regex
* Works for all projects, not just the first one in the workspace

tomkadwill commented on lib/provider.js:
```diff
@@ -1,7 +1,7 @@
 
 const STEP_DEFINITION_GLOB_CONFIG_KEY = 'cucumber-autocomplete-multipath.stepDefinitionGlob';
-const CUCUMBER_STEP_DEF_PATTERN = /(Given|And|When|Then)\(\/\^(.*?)\$/g;
+const CUCUMBER_STEP_DEF_PATTERN = /(Given|And|When|Then)\((?:'(.*?)'|\/\^(.*?)\$)/g;
```
This regex is becoming hard to read now :). I'd like to see some comments or some tests so that it is clear what it is matching

tomkadwill commented on lib/provider.js:
```diff
@@ -1,107 +1,57 @@
+      var pathGlob = [atom.config.get(STEP_DEFINITION_GLOB_CONFIG_KEY)];
+      var projectPath = atom.project.relativizePath(editor.getPath())[0];
+      var suggestions = [];
+      atom.workspace.scan(CUCUMBER_STEP_DEF_PATTERN, { paths: pathGlob }, function(match) {
```
👍

tomkadwill commented:
Hi @awilkins, sorry for the delay. This PR looks great! Really appreciate your work. I'm going to test this locally on my machine tomorrow and review your PR in more depth

tomkadwill commented:
Hey @awilkins, I just tried to run your branch on my machine and i'm getting this error:
```
/Applications/Atom.app/Contents/Resources/app.asar/src/task.js:52 Cannot read property 'split' of null TypeError: Cannot read property 'split' of null
    at processPaths (/Applications/Atom.app/Contents/Resources/app.asar/src/scan-handler.js:62:27)
    at /Applications/Atom.app/Contents/Resources/app.asar/src/scan-handler.js:36:21
    at /Applications/Atom.app/Contents/Resources/app.asar/node_modules/async/lib/async.js:99:13
    at Array.forEach (native)
    at _each (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/async/lib/async.js:32:24)
    at Object.async.each (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/async/lib/async.js:98:9)
    at Object.module.exports (/Applications/Atom.app/Contents/Resources/app.asar/src/scan-handler.js:33:18)
    at process.<anonymous> (/Applications/Atom.app/Contents/Resources/app.asar/src/task-bootstrap.js:84:10)
    at emitTwo (events.js:106:13)
    at process.emit (events.js:191:7)
    at process.nextTick (internal/child_process.js:744:12)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)(anonymous function) @ /Applications/Atom.app/Contents/Resources/app.asar/src/task.js:52
```

Any ideas?

awilkins commented:
Digging through this bug now...

Looks like it's really at line 48 [here](https://github.com/atom/atom/blob/1.14-releases/src/scan-handler.coffee)

The stack above this looks like this -

* workspace.scan sets up a bunch of DefaultDirectorySearcher instances
* They call scan-handler

`null` is an assigned value, so something is putting a null value in  `options.inclusions`

`scan()` sets up this property on line 1014 of `workspace.coffee`

Aha. A light beginneth to dawn.

In order not to clash with your package name I renamed it.

Then I changed it back because I wanted to do a PR. But I forgot to change the package name in the config constant.

Ergo, I'm passing a null into the glob array because I'm depending on the default value in the package definition in main.js to be present, but it's not there because the package name in package.json no longer matches the folder name I have the package in. Which is also presumably why the settings panel isn't showing up for me.

So it's all my fault. Branch updated.

Also removed the unit tests, which is obviously not perfect, but the code is completely different... would probably need a little refactoring to be more testable.

tomkadwill commented:
Thanks for your contribution @awilkins! I'll try to make some time this weekend to test this code out on my machine and give it a full review

tomkadwill commented:
Hey @awilkins, I can't seem to get this working on my machine :(. I'm not getting any errors but the package isn't working. Also I can't merge this because all the tests have been removed. 

Can I ask that you split this into multiple PRs for each thing. For example, the 'add linting' commit can be 1 PR. The 'replace config' commit can be another commit. That way it will be a lot easier for me to merge.

I'll keep this PR open for now, as a reference.

Thanks again for your work so far!
